### PR TITLE
Issue #2026 Retry Subworkflow Mechanism

### PR DESCRIPTION
**Pull request by mehmet-islamoglu**
Issue #2026: 

The requested changes from the previous Pull Request are made: 

Pull Request #2025 Fixed the retry mechanism for subworkflows so that retry starts from the last failed task. 

co-authors: @grkn @azb96

Please guide through the contribution process @kishorebanala @apanicker-nflx @aravindanr @jfernandez

### CODE REVIEW

**Comment by mehmet-islamoglu** [comment]
Our build is failing with this error. Can you help us please? 

@kishorebanala @apanicker-nflx @aravindanr @jfernandez

```
toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
No stopped containers
> Task :conductor-mysql-persistence:integrationTestDockerComposeUp FAILED
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':conductor-mysql-persistence:integrationTestDockerComposeUp'.
> Exit-code 1 when calling docker-compose, stdout: N/A
```

**Comment by mehmet-islamoglu** [comment]
@kishorebanala could you please review our PR?

File: jersey/src/main/java/com/netflix/conductor/server/resources/WorkflowResource.java [hunk 1]
```diff
@@ -189,8 +189,8 @@ public void restart(@PathParam("workflowId") String workflowId, @QueryParam("use
     @Path("/{workflowId}/retry")
     @ApiOperation("Retries the last failed task")
     @Consumes(MediaType.WILDCARD)
-    public void retry(@PathParam("workflowId") String workflowId) {
-        workflowService.retryWorkflow(workflowId);
+    public void retry(@PathParam("workflowId") String workflowId,@QueryParam("retryFromLastFailedTask") boolean retryFromLastFailedTask) {
```
Comment (kishorebanala) [inline jersey/src/main/java/com/netflix/conductor/server/resources/WorkflowResource.java:192]: Naming the parameter as `retryFromLastFailedTask` doesn't seem intuitive. For example, if the workflow doesn't have any subworkflow(s), the default behavior is to retry from last failed task. Setting this parameter `retryFromLastFailedTask` to `false` on retry request would then have no impact.

Instead, how about calling this `resumeSubworkflowTasks` or something similar?

Also, please add `@DefaultValue("false")` to QueryParam (See `restart` definition above for reference).

File: ui/src/components/workflow/executions/WorkflowAction.js [hunk 2]
```diff
@@ -65,6 +69,12 @@ class WorkflowAction extends React.Component {
       </Popover>
     );
 
+    const ttRetryLastFailedSub = (
+        <Popover id="popover-trigger-hover-focus" title="Retry Last Failed Task from Subworkflow">
```
Comment (kishorebanala) [inline ui/src/components/workflow/executions/WorkflowAction.js:73]: +1

File: test-harness/src/test/java/com/netflix/conductor/tests/integration/AbstractWorkflowServiceTest.java [hunk 3]
```diff
@@ -5209,6 +5209,158 @@ public void testSubWorkflowRetry() {
...
+        assertNotNull("Output: " + task.getOutputData().toString() + ", status: " + task.getStatus(), task.getSubWorkflowId());
+        String subWorkflowId = task.getSubWorkflowId();
+
+        workflow = workflowExecutionService.getExecutionStatus(subWorkflowId, true);
+        assertNotNull(workflow);
+        assertNotNull(workflow.getTasks());
+        assertEquals(workflowId, workflow.getParentWorkflowId());
+        assertEquals(RUNNING, workflow.getStatus());
+
+        // poll and fail the first task in sub-workflow
```
Comment (kishorebanala) [inline test-harness/src/test/java/com/netflix/conductor/tests/integration/AbstractWorkflowServiceTest.java:5276]: Comment doesn't match with the action.

**Review by kishorebanala — CHANGES_REQUESTED** [review]

**Comment by kishorebanala** [comment]
@mehmet-islamoglu @grkn @azb96 Great work so far, thanks for the contribution.

Regarding your question about progressing the PR, let's finalize the review and we'll take it from there to merge it. In general, after merging to dev, we batch merge dev to master and cut out a new release tag.

**Comment by mehmet-islamoglu** [comment]
Hello @kishorebanala,

We have completed all of your requested changes. Could you please review our PR again?

**Review by kishorebanala — APPROVED** [review]

**Pull request merged**